llvm/lib/Target/AMDGPU/GCNHazardRecognizer.cpp

```diff
@@ -875,13 +875,34 @@ GCNHazardRecognizer::checkVALUHazardsHelper(const MachineOperand &Def,
     return DataIdx >= 0 &&
            TRI->regsOverlap(MI.getOperand(DataIdx).getReg(), Reg);
   };
+
   int WaitStatesNeededForDef =
     VALUWaitStates - getWaitStatesSince(IsHazardFn, VALUWaitStates);
   WaitStatesNeeded = std::max(WaitStatesNeeded, WaitStatesNeededForDef);
 
   return WaitStatesNeeded;
 }
 
+static const MachineOperand *
+getDstSelForwardingOperand(const MachineInstr &MI, const GCNSubtarget &ST) {
+  if (!SIInstrInfo::isVALU(MI))
+    return nullptr;
+
+  const SIInstrInfo *TII = ST.getInstrInfo();
+  if (SIInstrInfo::isSDWA(MI)) {
+    if (auto *DstSel = TII->getNamedOperand(MI, AMDGPU::OpName::dst_sel))
+      if (DstSel->getImm() == AMDGPU::SDWA::DWORD)
+        return nullptr;
+  } else {
+    if (!AMDGPU::hasNamedOperand(MI.getOpcode(), AMDGPU::OpName::op_sel) ||
+        !(TII->getNamedOperand(MI, AMDGPU::OpName::src0_modifiers)->getImm() &
+          SISrcMods::DST_OP_SEL))
+      return nullptr;
+  }
```

Comment:
Two pieces to this hazard, the instruction that has dest forwarding and the consumer of the fowarded dest.

Different types of instructions with dest forwarding:

f1. SDWA w/ dest_sel
f2. VOP3 with op_sel[3]
f3. CVR_SR_FP8_F32 || CVR_SR_BF8_F32 w/ op_sel[2:3] != 0


And there are three classes of consumers:

c1. RAW
c2a. WAW (dest reg)
c2b. WAW (ECC)

c2a is sdwa w/ UNUSED_PRESERVE semantics. This results in incorrect bits in dest reg when the second partial write does not overlap completely with the first.

c2b is ecc only which is always enabled for MI300. For parity check, ecc will read the dst for compare after modifying the value. This occurs when we partial write to the same dest, and we preserve the initial value. For example, VOP3A 16 bit instructions with op_sel[3] have dest preserve semantics. However, VOP2 16 bit instructions (e.g.) don't have dest preserve semantics v_add_u16 so ecc isn't relevant. AFAICT, this only effects result of ecc.


"should it just consider any access to the register as the write"? For MI300 I don't think this is necessary. The trickiest is c2b. For MI300, VOP3A and all the loads with dest preserve (thanks for pointing this out) should be sufficient to represent the class. That said, maybe the better approach is to simply whitelist VOP1 and VOP2 WAW and insert s_nop for all others instead of trying to encode all these conditions.

Reply:
As per ISA Manual of Mi300 _(section 7.2)_, Irrespective of  _OPSEL_ value for CVT_*_F32, we need _nop_ if SRC of a VALU instruction is using result produced by previous CVT_*_F32 instructions.

Reply:
Thanks Pravin -- the definition of f1+f2+f3 is coming from MI300_SP_MAS section 1.3.9.2 which clearly states the op_sel requirements. I will run some more experiments and reach out to HW team to reconcile the conflicting documentation if necessary.

Reply:
I ran tests on MI300 with instructions in the following table. In version 1, each instruction was immediately followed by VALU RAW, in version 2, each instruction was immediately followed by s_nop then VALU RAW. Comparing these results to the expected value, I was able to determine if s_nop is needed:

| Instruction       | op_sel      | need s_nop |
| ----------------- | ----------- | ---------- |
| v_cvt_sr_fp8_f32  | [1,1,1,1]   |     yes    |
| v_cvt_sr_fp8_f32  | [1,1,1,0]   |     yes    |
| v_cvt_sr_fp8_f32  | [1,1,0,1]   |     yes    |
| v_cvt_sr_fp8_f32  | [1,1,0,0]   |     no     |
| v_cvt_sr_fp8_f32  | unspecified |     no     |
| v_cvt_pk_fp8_f32  | [1,1,1]     |     yes    |
| v_cvt_pk_fp8_f32  | [1,1,0]     |     no     |
| v_cvt_pk_fp8_f32  | unspecified |     no     |

v_cvt_sr_fp8_f32 op_sel requirements for dest forwarding issue are consistent with MI300_SP_MAS section 1.3.9.2  (op_sel[3:2] != 0)

v_cvt_pk_fp8_f32 op_sel requirements for dest forwarding issue are inconsistent with MI300_SP_MAS section 1.3.9.2 . Document states "VOP3 opcodes with op_sel[3] = 1 [don't have 4 cycle feedback]", but the control for dest bits for the instruction is op_sel[2]. op_sel[2] = 1 is the requirement for dest forwarding issue. Our TII calls already handle this situation as expected.

It seems there is only a dest forwarding issue if the instruction 1 write is not aligned to the register. If it is a partial write but aligned to the register then no swizzle / pack is needed and no dest forwarding issue.

Reply:
> However, VOP2 16 bit instructions (e.g. v_add_u16) don't have dest preserve semantics so ecc isn't relevant. 

This is more complicated than that, they made this a big mess. gfx8 always zeroed the top bits, gfx9 made it complicated, so some opcodes do, some don't, and some depend on op_sel. gfx10 made everything preserve. zeroesHigh16BitsOfDest is incomplete but theoretically should document which cases preserve or not. I think for example v_add_u16 in the vop3 form supports op_sel for the partial write behavior?

Reply:
Yes, `V_ADD_U16` doesn't preserve dest in VOP2, but can in VOP3. I thought the problem was limited to VOP3 due to op_sel

Thanks for pointing out that list..

I took another look at the ISA, and you're right.. there are VOP2 which also have dest preserve semantics -- e.g. V_MAC_F16 

So I guess we should safely s_nop any VALU WAW, and "optimize" once we can trust zeroesHigh16BitsOfDest